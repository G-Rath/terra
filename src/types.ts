import { AwsResourceType } from '@src/utils';

export enum TFNodeType {
  Resource = 'resource',
  Module = 'module',
  Dynamic = 'dynamic',
  Argument = 'argument',
  Function = 'function',
  Block = 'block',
  Map = 'map'
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface TFBaseNode {
  //
}

export type TFPrimitiveLiteral =
  | string //
  | number
  | boolean
  | null;

export interface TFMapLiteral extends TFBaseNode {
  type: TFNodeType.Map;
  attributes: Array<[string, TFLiteralExpression]>;
}

export type TFLiteralExpression =
  | TFPrimitiveLiteral // todo: replace w/ Node
  | TFFunctionCall
  | TFLiteralExpression[] // todo: replace w/ Node
  | TFMapLiteral;

interface TFFunctionCall extends TFBaseNode {
  type: TFNodeType.Function;
  name: string;
  args: unknown[];
}

export interface TFArgument<TIdentifier extends string = string>
  extends TFBaseNode {
  type: TFNodeType.Argument;
  identifier: TIdentifier;
  expression: TFLiteralExpression;
}

export interface TFBlockLiteral<TIdentifier extends string = string>
  extends TFBaseNode {
  type: TFNodeType.Block;

  /**
   * The name of the block literal
   */
  name: string;
  /**
   * The arguments that make up the resource.
   */
  body: TFBlockBody<TIdentifier>;
}

export interface TFDynamicBlock<TIdentifier extends string = string>
  extends TFBaseNode {
  type: TFNodeType.Dynamic;
  /**
   * The label of the dynamic block, which specifies what kind of nested block to generate.
   */
  name: string;
  content: TFBlockBody<TIdentifier>;
  labels: string[];
  forEach: unknown;
  /**
   * Sets the name of a temporary variable that represents the current element of the complex value.
   * If omitted, the name of the variable defaults to the label of the dynamic block.
   */
  iterator?: string;
}

export type TFBlockBodyElement<TIdentifier extends string = string> =
  | TFArgument<TIdentifier>
  | TFBlockLiteral
  | TFDynamicBlock;

export type TFBlockBody<TIdentifier extends string = string> = Array<
  TFBlockBodyElement<TIdentifier>
>;

export enum RandomResourceType {
  RandomString = 'random_string',
  RandomPet = 'random_pet',
  RandomId = 'random_id'
}

export type ResourceType =
  | AwsResourceType //
  | RandomResourceType
  | string;

export type TFTopLevelBlock = TFResourceBlock | TFModuleBlock;

/** The AST for a valid Terraform file. */
export type TFFileAST = TFTopLevelBlock[];

/**
 * A Terraform module, made up of ASTs of the files in the module.
 * Note that a "module" is different from a "module block".
 */
interface TFModule {
  files: { [K: string]: TFFileAST };
}

export interface TFResourceBlock<TIdentifier extends string = string> {
  type: TFNodeType.Resource;
  /**
   * The type of this resource.
   */
  resource: ResourceType;
  /**
   * The name of this resource.
   */
  name: string;
  /**
   * The body of this resource block.
   */
  body: TFBlockBody<TIdentifier>;
}

export interface TFModuleBlock<TIdentifier extends string = string>
  extends TFBaseNode {
  type: TFNodeType.Module;
  /**
   * The name of this module.
   */
  name: string;
  /**
   * The body of this module block.
   */
  body: TFBlockBody<TIdentifier | 'source'>;
}
