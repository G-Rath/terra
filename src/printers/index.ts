export * from './printBlockBody';
export * from './printBlockLiteral';
export * from './printFileAST';
export * from './printResourceBlock';
