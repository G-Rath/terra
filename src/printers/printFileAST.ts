import { printTFBlock } from '@src/printer';
import { TFFileAST } from '@src/types';

export const printFileAST = (fileAST: TFFileAST) =>
  fileAST.map(printTFBlock).join('\n\n');
