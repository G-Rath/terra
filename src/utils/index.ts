export * from './AwsResourceType';
export * from './asResourceName';
export * from './makeTFArgument';
export * from './makeTFBlockLiteral';
export * from './makeTFDynamicBlock';
export * from './makeTFResourceBlock';
export * from './makeTFStringArgument';
export * from './normaliseRoute53Name';
