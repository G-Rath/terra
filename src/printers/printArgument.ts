import { printLiteralExpression } from '@src/printers';
import { TFArgument } from '@src/types';

export const printArgument = (blockArgument: TFArgument): string =>
  [
    blockArgument.identifier,
    '=',
    printLiteralExpression(blockArgument.expression)
  ].join(' ');
