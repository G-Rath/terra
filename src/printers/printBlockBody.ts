import { printArgument, printBlockLiteral } from '@src/printers';
import { TFBlockBody, TFNodeType } from '@src/types';
import indentString from 'indent-string';

export const printBlockBody = (body: TFBlockBody): string => {
  if (body.length === 0) {
    return '{}';
  }

  return [
    '{',
    ...body
      .map(node => {
        if (node.type === TFNodeType.Argument) {
          return printArgument(node);
        }

        if (node.type === TFNodeType.Block) {
          return printBlockLiteral(node);
        }

        if (node.type === TFNodeType.Dynamic) {
          return `# FIXME: ${node.type} is not yet supported`;
        }

        throw new Error(`structural error: unknown node type "${node}"`);
      })
      .map(str => indentString(str, 2)),
    '}'
  ].join('\n');
};
