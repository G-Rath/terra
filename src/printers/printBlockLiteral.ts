import { printBlockBody } from '@src/printers';
import { TFBlockLiteral } from '@src/types';

/**
 * Prints the given `blockLiteral`
 *
 * @param {TFBlockLiteral} blockLiteral
 *
 * @return {string}
 */
export const printBlockLiteral = (blockLiteral: TFBlockLiteral): string =>
  [blockLiteral.name, printBlockBody(blockLiteral.body.body)].join(' ');
