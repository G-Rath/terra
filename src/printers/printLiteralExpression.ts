import { printPrimitiveLiteral } from '@src/printers';
import { TFLiteralExpression, TFNodeType } from '@src/types';
import indentString from 'indent-string';

export const printLiteralExpression = (
  literal: TFLiteralExpression
): string => {
  if (literal === null || typeof literal !== 'object') {
    return printPrimitiveLiteral(literal);
  }

  if (Array.isArray(literal)) {
    return [
      '[',
      indentString(
        literal
          .map(subexpression => `${printLiteralExpression(subexpression)}`)
          .join(',\n'),
        2
      ),
      ']'
    ].join('\n');
  }

  if (literal.type === TFNodeType.Map) {
    return [
      '{',
      ...literal.attributes
        .map(([identifier, expression]) =>
          [identifier, '=', printLiteralExpression(expression)].join(' ')
        )
        .map(str => indentString(str, 2)),
      '}'
    ].join('\n');
  }

  if (literal.type === TFNodeType.Function) {
    return [
      `${literal.name}(`,
      ...(['# FIXME - FUNCTIONS NOT YET SUPPORTED'] || literal.args).map(str =>
        indentString(str, 2)
      ),
      ')'
    ].join('\n');
  }

  throw new Error('structural error');
};
