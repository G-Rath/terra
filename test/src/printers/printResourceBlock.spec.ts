import {
  makeTFArgument,
  makeTFBlockLiteral,
  makeTFResourceBlock
} from '@src/makers';
import { printResourceBlock } from '@src/printers';
import { TFNodeType, TFResourceBlock } from '@src/types';
import { AwsResourceType } from '@src/utils';

describe('printResourceBlock', () => {
  it.each<TFResourceBlock>([
    makeTFResourceBlock('imnotcrazy_info', AwsResourceType.AWS_ROUTE53_ZONE, [
      makeTFArgument('name', '"imnotcrazy.info"'),
      makeTFArgument('tags', {
        type: TFNodeType.Function,
        name: 'merge',
        args: ['local.common_tags']
      })
    ]),
    makeTFResourceBlock(
      'imnotcrazy_info_ns',
      AwsResourceType.AWS_ROUTE53_RECORD,
      [
        ...([
          ['allow_overwrite', true],
          ['name', '""'],
          ['ttl', 300],
          ['type', '"NS"'],
          ['zone_id', 'aws_route53_zone.imnotcrazy_info.zone_id']
        ] as const).map(([identifier, expression]) =>
          makeTFArgument(identifier, expression)
        ),
        makeTFArgument('records', [
          'aws_route53_zone.imnotcrazy_info.name_servers.0',
          'aws_route53_zone.imnotcrazy_info.name_servers.1',
          'aws_route53_zone.imnotcrazy_info.name_servers.2',
          'aws_route53_zone.imnotcrazy_info.name_servers.3'
        ])
      ]
    ),
    makeTFResourceBlock(
      'distribution',
      AwsResourceType.AWS_CLOUDFRONT_DISTRIBUTION,
      [
        makeTFArgument('allowed_methods', ['"GET"', '"HEAD"']),
        makeTFArgument('cached_methods', ['"GET"', '"HEAD"']),
        ...([
          ['compress', true],
          ['default_ttl', 31536000],
          ['max_ttl', 31536000],
          ['min_ttl', 0],
          ['path_pattern', '"/static/*"'],
          ['smooth_streaming', false],
          ['target_origin_id', '"S3-app.mine/test"'],
          ['viewer_protocol_policy', '"https-only"']
        ] as const).map(([identifier, expression]) =>
          makeTFArgument(identifier, expression)
        ),
        makeTFArgument('trusted_signers', []),
        makeTFBlockLiteral('forwarded_values', [
          makeTFArgument('headers', []),
          makeTFArgument('query_string', false),
          makeTFArgument('query_string_cache_keys', []),
          makeTFBlockLiteral('cookies', [
            makeTFArgument('forward', '"none"'),
            makeTFArgument('whitelisted_names', [])
          ])
        ])
      ]
    )
  ])('prints as expected', resourceBlock => {
    expect(printResourceBlock(resourceBlock)).toMatchSnapshot();
  });
});
