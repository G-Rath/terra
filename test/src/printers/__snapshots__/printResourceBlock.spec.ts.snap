// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printResourceBlock prints as expected 1`] = `
"resource \\"aws_route53_zone\\" \\"imnotcrazy_info\\" {
  name = \\"imnotcrazy.info\\"
  tags = merge(
    # FIXME - FUNCTIONS NOT YET SUPPORTED
  )
}"
`;

exports[`printResourceBlock prints as expected 2`] = `
"resource \\"aws_route53_record\\" \\"imnotcrazy_info_ns\\" {
  allow_overwrite = true
  name = \\"\\"
  ttl = 300
  type = \\"NS\\"
  zone_id = aws_route53_zone.imnotcrazy_info.zone_id
  records = [
    aws_route53_zone.imnotcrazy_info.name_servers.0,
    aws_route53_zone.imnotcrazy_info.name_servers.1,
    aws_route53_zone.imnotcrazy_info.name_servers.2,
    aws_route53_zone.imnotcrazy_info.name_servers.3
  ]
}"
`;

exports[`printResourceBlock prints as expected 3`] = `
"resource \\"aws_cloudfront_distribution\\" \\"distribution\\" {
  allowed_methods = [
    \\"GET\\",
    \\"HEAD\\"
  ]
  cached_methods = [
    \\"GET\\",
    \\"HEAD\\"
  ]
  compress = true
  default_ttl = 31536000
  max_ttl = 31536000
  min_ttl = 0
  path_pattern = \\"/static/*\\"
  smooth_streaming = false
  target_origin_id = \\"S3-app.mine/test\\"
  viewer_protocol_policy = \\"https-only\\"
  trusted_signers = [

  ]
  forwarded_values {
    headers = [

    ]
    query_string = false
    query_string_cache_keys = [

    ]
    cookies {
      forward = \\"none\\"
      whitelisted_names = [

      ]
    }
  }
}"
`;
